eox_tagging/test/test_models.py

```diff
@@ -0,0 +1,183 @@
+
+    def test_tag_type(self):
+        """ Used to confirm that the tag_type is correct."""
+        tag_value = getattr(self.test_tag, "tag_type")
+
+        self.assertEqual(tag_value, "testType")
+
+    def test_tag_value_not_in_settings(self):
+        """ Used to confirm validation error when the value is not defined."""
+        with self.assertRaises(ValidationError):
+            Tag.objects.create(
+                tag_value="testValues",
+                tag_type="testType",
+                tagged_object=self.tagged_object,
+                belongs_to=self.belongs_to_object,
+            )
+
+    def test_tag_type_not_in_settings(self):
+        """ Used to confirm validation error when the type is not defined."""
+        with self.assertRaises(ValidationError):
+            Tag.objects.create(
+                tag_value="testValue",
+                tag_type="testTypes",
+                tagged_object=self.tagged_object,
+                belongs_to=self.belongs_to_object,
+            )
+
+    def test_tag_different_generic_objects_fail(self):
+        """ Used to confirm the tags can't be use with objects not defined in settings."""
+        with self.assertRaises(ValidationError):
+            Tag.objects.create(
+                tag_value="testValue",
+                tag_type="testType",
+                tagged_object=self.fake_object_tagged_fail,
+                belongs_to=self.belongs_to_object,
+            )
+
+    def test_tag_different_generic_objects(self):
+        """ Used to confirm the tags can be use with any object defined in settings."""
+        Tag.objects.create(
+            tag_value="testValue",
+            tag_type="testType",
+            tagged_object=self.fake_object_tagged,
+            belongs_to=self.belongs_to_object,
+        )
+
+    @override_settings(EOX_TAGGING_DEFINITIONS=[
+        {"field_name": "tag_value", "validations": ["definition"], "allowed": r".*"},
+        {"field_name": "tagged_object_name", "validations": ["definition"], "allowed": r".*"}])
+    def test_tag_validation_regex_accepts_any_value(self):
+        """ Used to confirm that tags can match any string if defined in settings."""
+        Tag.objects.create(
+            tag_value="anything can be here",
+            tag_type="testType",
+            tagged_object=self.tagged_object,
+            belongs_to=self.belongs_to_object,
+        )
+
+        Tag.objects.create(
+            tag_value="anything can be here",
+            tag_type="testType",
+            tagged_object=self.fake_object_tagged_fail,  # In previus tests failed, not in this
+            belongs_to=self.belongs_to_object,
+        )
+
+    @override_settings(EOX_TAGGING_DEFINITIONS=[
+        {"field_name": "tag_value", "validations": ["definition"], "allowed": r"^a...s$"}])
+    def test_tag_validation_regex_accepts_pattern(self):
+        """ Used to confirm that tags can accept a pattern if defined in settings."""
+        Tag.objects.create(
```

Comment:
x3

Reply:
For example: 
With this setting, the field tag_value has a validation defined, called _definition_. This validation means that the field tag_value must follow the definition described in _allowed_. In this case, it must match the pattern.
What we are testing is that we can create the object that follows that definition. The test _test_tag_validation_regex_accepts_pattern_fail_ demonstrates that if the field does not follow the definition, then the creation fails